newIDE/app/src/EventsFunctionsExtensionsLoader/index.js

```diff
@@ -484,98 +556,116 @@ function generateBehavior(
+      behaviorMethodMangledNames.set(
         eventsFunction.getName(),
         eventsFunctionMangledName
       );
+    }
 
-      const instructionOrExpression = declareObjectInstructionOrExpressionMetadata(
-        extension,
-        objectMetadata,
-        eventsBasedObject,
-        eventsFunction
-      );
-      declareEventsFunctionParameters(
-        eventsFunctionsContainer,
-        eventsFunction,
-        instructionOrExpression,
-        1
-      );
+    const instructionOrExpression = declareBehaviorInstructionOrExpressionMetadata(
+      extension,
+      behaviorMetadata,
+      eventsBasedBehavior,
+      eventsFunction
+    );
+    declareEventsFunctionParameters(
+      eventsFunctionsContainer,
+      eventsFunction,
+      instructionOrExpression,
+      2
+    );
 
-      // Hide "lifecycle" methods as they are called automatically by
-      // the game engine.
-      if (isObjectLifecycleEventsFunction(eventsFunction.getName())) {
-        instructionOrExpression.setHidden();
-      }
+    // Hide "lifecycle" methods as they are called automatically by
+    // the game engine.
+    if (isBehaviorLifecycleEventsFunction(eventsFunction.getName())) {
+      instructionOrExpression.setHidden();
+    }
 
-      if (eventsFunction.isPrivate()) instructionOrExpression.setPrivate();
+    if (eventsFunction.isPrivate()) instructionOrExpression.setPrivate();
 
-      instructionOrExpression.setIncludeFile(includeFile);
+    instructionOrExpression.setIncludeFile(includeFile);
 
-      if (
-        eventsFunction.isAsync() &&
-        typeof instructionOrExpression.setAsyncFunctionName === 'function'
-      )
-        //$FlowFixMe
-        instructionOrExpression.setAsyncFunctionName(eventsFunctionMangledName);
-      else instructionOrExpression.setFunctionName(eventsFunctionMangledName);
-    });
+    if (
+      eventsFunction.isAsync() &&
+      typeof instructionOrExpression.setAsyncFunctionName === 'function'
+    )
+      //$FlowFixMe
+      instructionOrExpression.setAsyncFunctionName(eventsFunctionMangledName);
+    else instructionOrExpression.setFunctionName(eventsFunctionMangledName);
+  });
+  return behaviorMetadata;
+}
+
+function generateObject(
+  project: gdProject,
+  extension: gdPlatformExtension,
+  eventsFunctionsExtension: gdEventsFunctionsExtension,
+  eventsBasedObject: gdEventsBasedObject,
+  options: OptionsForGeneration,
+  codeGenerationContext: CodeGenerationContext
+): Promise<void> {
+  return Promise.resolve().then(() => {
+    const objectMethodMangledNames = new gd.MapStringString();
+    const objectMetadata = generateObjectMetadata(
+      project,
+      extension,
+      eventsFunctionsExtension,
+      eventsBasedObject,
+      options,
+      codeGenerationContext,
+      objectMethodMangledNames
+    );
 
     // Generate code for the object and its methods
     if (!options.skipCodeGeneration) {
+      const codeNamespace = getObjectFunctionCodeNamespace(
```

Comment:
You don't need the promise.resolve().then() if you use `async`, which takes care for you of returning a promise even if you've not awaited/returned a promise (it will wrap your return value in a Promise.resolve)

```suggestion
async function generateObject(
  project: gdProject,
  extension: gdPlatformExtension,
  eventsFunctionsExtension: gdEventsFunctionsExtension,
  eventsBasedObject: gdEventsBasedObject,
  options: OptionsForGeneration,
  codeGenerationContext: CodeGenerationContext
): Promise<void> {

    const objectMethodMangledNames = new gd.MapStringString();
    const objectMetadata = generateObjectMetadata(
      project,
      extension,
      eventsFunctionsExtension,
      eventsBasedObject,
      options,
      codeGenerationContext,
      objectMethodMangledNames
    );

    // Generate code for the object and its methods
    if (!options.skipCodeGeneration) {
      const codeNamespace = getObjectFunctionCodeNamespace(
```

Reply:
I didn't modified this code. Do you think I should convert the call tree to async?